cs.c

```diff
@@ -1156,85 +1224,80 @@ size_t CAPSTONE_API cs_disasm(csh ud, const uint8_t *buffer, size_t size, uint64
 
 	handle->errnum = CS_ERR_OK;
 
-#ifdef CAPSTONE_USE_SYS_DYN_MEM
-	if (count > 0 && count <= INSN_CACHE_SIZE)
-		cache_size = (unsigned int) count;
-#endif
+	cs_buffer_clear(buffer);
 
 	// save the original offset for SKIPDATA
-	buffer_org = buffer;
-	offset_org = offset;
-	size_org = size;
-
-	total_size = sizeof(cs_insn) * cache_size;
-	total = cs_mem_calloc(sizeof(cs_insn), cache_size);
-	if (total == NULL) {
-		// insufficient memory
-		handle->errnum = CS_ERR_MEM;
-		return 0;
-	}
+	code_org = code;
+	code_size_org = code_size;
+	address_org = address;
 
-	insn_cache = total;
+	for (; code_size && (count == 0 || buffer->count < count); ++buffer->count) {
+		if (buffer->capacity <= buffer->count && !cs_buffer_reserve(buffer, 1)) {
```

Comment:
Couldn't you also reserve the buffer once, on based on some good guess how many instructions will be disassembled? So most often `code_size / 4`, sometimes `code_size / 2`.

Reply:
Good idea!